upstream/upstream.go

```diff
@@ -20,17 +25,25 @@ type Backend interface {
 	HealthCheck(ctx context.Context) error
```

Comment:
instead of having `DisableTiers`, we could have two interfaces for the backend, and choose tiered/not-tiered approach based on the interface:

```golang
HealthCheck(context.Context) error
```

```golang
type Tier int
HealthCheckTiered(context.Context) (Tier, error)
```

this way if `HealthCheck` is implemented, `Tier` is assumed to be zero, and all nodes are of the same tier, and it works like before.

but if the `HealthCheckTiered` is implemented, `Tier` becomes dynamic (and we do StableSort on it)

we probably can simply sort on `Tier` ignoring `Score`, this way the order of the nodes won't change.

we can keep new backends with `Tier` of zero before we have the healthcheck

this way we also allow `Tier` calculation to be customized - e.g. some upstreams might have fixed tier no matter what is the latency, etc.

Reply:
That is doable, but I will **emphasize** here that we are moving any guarantees about Backend tier and proper placement from List internals to the outside. That may not a be bad thing, in case we trust `HealthCheckTiered` to be implemented currently. But for me, it's a bit strange that we manage balancer internal state from, essentially, outside.